pkg/governance-scripts/contracts/TimelockAuthorizerTransitionMigrator.sol

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/vault/IBasicAuthorizer.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-vault/contracts/authorizer/TimelockAuthorizer.sol";
+
+/**
+ * @notice Migrates permissions granted after `TimelockAuthorizer` deployment from the old authorizer.
+ */
+contract TimelockAuthorizerTransitionMigrator {
+    using Address for address;
+
+    TimelockAuthorizer public immutable timelockAuthorizer;
+    RoleData[] public rolesData;
+
+    struct RoleData {
+        address grantee;
+        bytes32 role;
+        address target;
+    }
+
+    bool private _migrationDone;
+
+    /**
+     * @dev Reverts if rolesData contains a role for an account which doesn't hold the same role on the old Authorizer.
+     */
+    constructor(
+        IBasicAuthorizer oldAuthorizer,
+        TimelockAuthorizer _timelockAuthorizer,
+        RoleData[] memory _rolesData
+    ) {
+        timelockAuthorizer = _timelockAuthorizer;
+
+        for (uint256 i = 0; i < _rolesData.length; i++) {
+            RoleData memory roleData = _rolesData[i];
+            // We require that any permissions being copied from the old Authorizer must exist on the old Authorizer.
+            // This simplifies verification of the permissions being added to the new TimelockAuthorizer.
+            require(oldAuthorizer.canPerform(roleData.role, roleData.grantee, roleData.target), "UNEXPECTED_ROLE");
```

Comment:
It means you'd have to store the old authorizer address, but it would be safer to check this in `migratePermissions` - in case there is a delay between deployment and whenever `migratePermissions` is called, and any of the old permissions are revoked. (It wouldn't revert, just skip any that had been revoked, maybe emitting an event.) Granting permissions will already emit an event.

Reply:
Good point. I've added the check in `migratePermissions` in https://github.com/balancer-labs/balancer-v2-monorepo/pull/2150/commits/615be11bd84e8df3ed57cbaaccb7a15a19d684c6.
I'd keep this require anyways, since it's pointless to store permissions that are not present in the old authorizer at creation time -- it would mean we are doing something wrong during the deployment.